codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java

```diff
@@ -241,7 +241,18 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
                 ByteBuf data = (ByteBuf) msg;
                 first = cumulation == null;
                 if (first) {
-                    cumulation = data;
+                    if (data.isReadOnly()) {
+                        // If the ByteBuf is read-only we will do a memory copy as otherwise we can not write data
+                        // to it on the next channelRead(...) call.
+                        cumulation = ctx.alloc().buffer(data.readableBytes());
```

Comment:
Shouldn't this logic be part of a cumulator implementation? Some cumulator implementation (the one that uses a CompositeBytebuf) may not need a copy at all.